src/driver/xen.rs

```diff
@@ -300,71 +328,106 @@ impl Introspectable for Xen {
     }
 
     fn listen(&mut self, timeout: u32) -> Result<Option<Event>, Box<dyn Error>> {
-        let fd = self.xev.xenevtchn_fd()?;
-        let fd_struct = PollFd::new(fd, PollFlags::POLLIN | PollFlags::POLLERR);
-        let mut fds = [fd_struct];
-        let mut vcpu: u16 = 0;
-        let mut event_type = unsafe { mem::MaybeUninit::<EventType>::zeroed().assume_init() };
-        let poll_result = poll(&mut fds, timeout.try_into().unwrap()).unwrap();
-        let mut pending_event_port = -1;
-        if poll_result == 1 {
-            pending_event_port = self.xev.xenevtchn_pending()?;
-            if pending_event_port != -1 {
-                self.xev
-                    .xenevtchn_unmask(pending_event_port.try_into().unwrap())?;
+        let mut fds: [PollFd; 1] = [self.evtchn_pollfd];
+        let event: Option<Event> = match poll(&mut fds, timeout.try_into()?)? {
+            0 => {
+                // timeout. no file descriptors were ready
+                None
             }
-        }
-        let back_ring_ptr = &mut self.back_ring;
-        let mut flag = false;
-        if poll_result > 0
-            && self.xev.get_bind_port() == pending_event_port
-            && RING_HAS_UNCONSUMED_REQUESTS!(back_ring_ptr) != 0
-        {
-            flag = true;
-            let req = self.xc.get_request(back_ring_ptr)?;
-            if req.version != VM_EVENT_INTERFACE_VERSION {
-                panic!("version mismatch");
+            -1 => {
+                // failure
+                return Err(Box::new(IoError::last_os_error()));
             }
-            let xen_event_type = (self.xc.get_event_type(req)).unwrap();
-            event_type = match xen_event_type {
-                XenEventType::Cr { cr_type, new, old } => EventType::Cr {
-                    cr_type: match cr_type {
-                        XenCr::Cr0 => CrType::Cr0,
-                        XenCr::Cr3 => CrType::Cr3,
-                        XenCr::Cr4 => CrType::Cr4,
-                    },
-                    new,
-                    old,
-                },
-                XenEventType::Msr { msr_type, value } => EventType::Msr { msr_type, value },
-                XenEventType::Breakpoint { insn_len, .. } => {
-                    EventType::Breakpoint { gpa: 0, insn_len }
+            1 => {
+                // event available
+                match self.xev.xenevtchn_pending()? {
+                    -1 => {
+                        // no event channel port is pending
+                        // TODO: Err
```

Comment:
Is there a reason why you haven't implemented this yet? Feels to me like it should be done in this PR.

Reply:
Because this means creating custom error types, and I'm not feeling confident / knowleadgable enough on the topic to implement this.

Reply:
Yeah, the overall error handling in this library is not very good. I once tried to refactor it, but got discouraged after realizing how much I would need to change.

Reply:
At some point we should create a custom error type (or perhaps even multiple, one for each driver) and implement error wrapping instead of error boxing. This is the way. ;)

Reply:
I'm totally in favor of something like this.
I'm simply lacking the time to invest in learning state of the art Rust, and to reflect the changes in libmicrovmi.

But my heart breaks every time I call `unwrap()` in libmicrovmi simply because I'm a noob in error handling.

Reply:
:D